lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp

```diff
@@ -282,35 +387,92 @@ SILCombiner::visitUncheckedAddrCastInst(UncheckedAddrCastInst *UADCI) {
 }
 
 SILInstruction *
-SILCombiner::visitUncheckedRefCastInst(UncheckedRefCastInst *URCI) {
-  if (URCI->getFunction()->hasOwnership())
-    return nullptr;
+SILCombiner::visitUncheckedRefCastInst(UncheckedRefCastInst *urci) {
+  // %0 = unchecked_ref_cast %x : $X->Y
+  // %1 = unchecked_ref_cast %0 : $Y->Z
+  //
+  // ->
+  //
+  // %0 = unchecked_ref_cast %x : $X->Y
+  // %1 = unchecked_ref_cast %x : $X->Z
+  //
+  // NOTE: For owned values, we only perform this optimization if we can
+  // guarantee that we can eliminate the initial unchecked_ref_cast.
+  if (auto *otherURCI = dyn_cast<UncheckedRefCastInst>(urci->getOperand())) {
+    SILValue otherURCIOp = otherURCI->getOperand();
+    if (otherURCIOp.getOwnershipKind() != OwnershipKind::Owned) {
+      return Builder.createUncheckedRefCast(urci->getLoc(), otherURCIOp,
+                                            urci->getType());
+    }
+    SingleBlockOwnedForwardingInstFolder folder(*this, urci);
+    if (folder.add(otherURCI)) {
+      auto *newValue = Builder.createUncheckedRefCast(
+          urci->getLoc(), otherURCIOp, urci->getType());
+      return std::move(folder).optimizeWithReplacement(newValue);
```

Comment:
I'm confused about why you're creating a new instruction here.

Reply:
I am matching what the code did previously.